<a href="https://colab.research.google.com/github/vlx300/kb_colab/blob/master/Python_Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Python Pandas Data Exploration**
![alt text](https://strongautomotive.com/wp-content/uploads/2014/06/460637863-e1403800392554.jpg)

#**Mount Google Drive**

In [2]:
 from google.colab import drive  # mount google drive #
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


###**Append sys.path for Google Drive**

In [0]:
import sys
sys.path.append("/content/drive/My Drive")

##**Grabbing the dataset**: 
**Five-thirty-Eight: NBA Results**

In [0]:
import requests   # import requestas library

download_url = "https://raw.githubusercontent.com/fivethirtyeight/data/master/nba-elo/nbaallelo.csv"  #data source
target_csv_path = "/content/drive/My Drive/nba_all_elo.csv"

In [5]:
response = requests.get(download_url)
response.raise_for_status()  # check that the request was successful 
with open(target_csv_path, "wb") as f:
  f.write(response.content)
print("download ready")

download ready


#**Python Pandas: Taking a closer look at the Data**

as you ca see above, we import pandas as pd alias, then we utilize read_csv() to read in the dataset, and store as a Pandas dataframe in a varible called nba. 

**Note:** Is your data NOT cvs format? the Pandas libray contains several other similar functions (read_json(), read_html(), read_sql_table()

In [6]:
import pandas as pd
nba = pd.read_csv("/content/drive/My Drive/nba_all_elo.csv") # store as pandas dataframe object in  varoiable called nba
type(nba)


pandas.core.frame.DataFrame

In [7]:
nba.head()

gameorder       game_id lg_id  ...  game_result  forecast notes
0          1  194611010TRH   NBA  ...            L  0.640065   NaN
1          1  194611010TRH   NBA  ...            W  0.359935   NaN
2          2  194611020CHS   NBA  ...            W  0.631101   NaN
3          2  194611020CHS   NBA  ...            L  0.368899   NaN
4          3  194611020DTF   NBA  ...            L  0.640065   NaN

[5 rows x 23 columns]

You can utilize the built-in len() function to determine the numbr of rows

In [8]:
len(nba)  # how much data does nba contain

126314

In [9]:
nba.shape # you can see it dimensionality .TUPLE containing the number of rows and columns

(126314, 23)

Now we know the dimensions of our dataset, How can we be sure that the dataset actually contains basketball stats? Take a look a tthe 1st five rows with head()

In [10]:
nba.head()

gameorder       game_id lg_id  ...  game_result  forecast notes
0          1  194611010TRH   NBA  ...            L  0.640065   NaN
1          1  194611010TRH   NBA  ...            W  0.359935   NaN
2          2  194611020CHS   NBA  ...            W  0.631101   NaN
3          2  194611020CHS   NBA  ...            L  0.368899   NaN
4          3  194611020DTF   NBA  ...            L  0.640065   NaN

[5 rows x 23 columns]

You can configure pandas to display all columns like this

In [0]:
pd.set_option("display.max.columns", None)

Also change the number of decimal places to display.

In [0]:
pd.set_option("display.precision", 2)

In [13]:
nba.head()  # display first five rows

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   
2          2  194611020CHS   NBA        0     1947  11/2/1946           1   
3          2  194611020CHS   NBA        1     1947  11/2/1946           2   
4          3  194611020DTF   NBA        0     1947  11/2/1946           1   

   is_playoffs team_id  fran_id  pts    elo_i    elo_n  win_equiv opp_id  \
0            0     TRH  Huskies   66  1300.00  1293.28      40.29    NYK   
1            0     NYK   Knicks   68  1300.00  1306.72      41.71    TRH   
2            0     CHS    Stags   63  1300.00  1309.65      42.01    NYK   
3            0     NYK   Knicks   47  1306.72  1297.07      40.69    CHS   
4            0     DTF  Falcons   33  1300.00  1279.62      38.86    WSC   

   opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location game_result  \
0    Knicks       68    1300.00    1306.72             H           L   
1   Huskies       66    1300.00    1293.28             A           W   
2    Knicks       47    1306.72    1297.07             H           W   
3     Stags       63    1300.00    1309.65             A           L   
4  Capitols       50    1300.00    1320.38             H           L   

   forecast notes  
0      0.64   NaN  
1      0.36   NaN  
2      0.63   NaN  
3      0.37   NaN  
4      0.64   NaN

In [14]:
nba.tail()   # display last five rows

gameorder       game_id lg_id  _iscopy  year_id  date_game  \
126309      63155  201506110CLE   NBA        0     2015  6/11/2015   
126310      63156  201506140GSW   NBA        0     2015  6/14/2015   
126311      63156  201506140GSW   NBA        1     2015  6/14/2015   
126312      63157  201506170CLE   NBA        0     2015  6/16/2015   
126313      63157  201506170CLE   NBA        1     2015  6/16/2015   

        seasongame  is_playoffs team_id    fran_id  pts    elo_i    elo_n  \
126309         100            1     CLE  Cavaliers   82  1723.41  1704.39   
126310         102            1     GSW   Warriors  104  1809.98  1813.63   
126311         101            1     CLE  Cavaliers   91  1704.39  1700.74   
126312         102            1     CLE  Cavaliers   97  1700.74  1692.09   
126313         103            1     GSW   Warriors  105  1813.63  1822.29   

        win_equiv opp_id   opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
126309      60.31    GSW   Warriors      103    1790.96    1809.98   
126310      68.01    CLE  Cavaliers       91    1704.39    1700.74   
126311      60.01    GSW   Warriors      104    1809.98    1813.63   
126312      59.29    GSW   Warriors      105    1813.63    1822.29   
126313      68.52    CLE  Cavaliers       97    1700.74    1692.09   

       game_location game_result  forecast notes  
126309             H           L      0.55   NaN  
126310             H           W      0.77   NaN  
126311             A           L      0.23   NaN  
126312             H           L      0.48   NaN  
126313             A           W      0.52   NaN

##**Getting to know the data**
![alt text](https://miro.medium.com/max/403/1*TUw8sTRFiMgV0KRyUzjWmQ.jpeg)


you imported a CSV file, with the pandas libray and had a 1st look a the contents of your dataset, so far you only seen the size of the dataset and the first/last few rows. 

**Now lets examine the data more systematically**

**Displaying the Data Types**

The first steps in getting to know your data is having a understanding of the **data-types** it contains.  While you can place anything into a list, the columns of a **Dataframe** contain values of a specific data type. When you compare Pandas and Python data structures you see that this behavior makes Pandas much faster.  

you can display ALL columns and their data types with .info()

In [15]:
nba.info

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 126314 entries, 0 to 126313
Data columns (total 23 columns):
gameorder        126314 non-null int64
game_id          126314 non-null object
lg_id            126314 non-null object
_iscopy          126314 non-null int64
year_id          126314 non-null int64
date_game        126314 non-null object
seasongame       126314 non-null int64
is_playoffs      126314 non-null int64
team_id          126314 non-null object
fran_id          126314 non-null object
pts              126314 non-null int64
elo_i            126314 non-null float64
elo_n            126314 non-null float64
win_equiv        126314 non-null float64
opp_id           126314 non-null object
opp_fran         126314 non-null object
opp_pts          126314 non-null int64
opp_elo_i        126314 non-null float64
opp_elo_n        126314 non-null float64
game_location    126314 non-null object
game_result      126314 non-null object
forecast         126314 non-null float64
notes     

You see a list of ALL the columns in your dataset and the type of data each column contains.  You can see three (3)**bold text** data types: **int64, float64**, and **object**.  Pandas use NumPy library to work with these types Later you will the more complex "categorical" data type which the Pandas library implements itself. 


the **object** data type is a special one. The Object data type is a "*catch-all*" for columns that Pandas doesnt recognize as any pother specific data type.  In practice,  it means that the values in the columns are strings.

although you can store arbitrary python objects in the object data type, there are some drawbacks

*   Strange values in a object column can harm pandas performance
*   interoperaility with other libraries can also be affected. 



##**Showing Basic Statistics**

Now that you've seen what datatypes are in your dataset, lets get an overview of the values in each column. You can do this with .describe()

In [16]:
nba.describe()

gameorder   _iscopy    year_id  seasongame  is_playoffs        pts  \
count  126314.00  126314.0  126314.00   126314.00    126314.00  126314.00   
mean    31579.00       0.5    1988.20       43.53         0.06     102.73   
std     18231.93       0.5      17.58       25.38         0.24      14.81   
min         1.00       0.0    1947.00        1.00         0.00       0.00   
25%     15790.00       0.0    1975.00       22.00         0.00      93.00   
50%     31579.00       0.5    1990.00       43.00         0.00     103.00   
75%     47368.00       1.0    2003.00       65.00         0.00     112.00   
max     63157.00       1.0    2015.00      108.00         1.00     186.00   

           elo_i      elo_n  win_equiv    opp_pts  opp_elo_i  opp_elo_n  \
count  126314.00  126314.00  126314.00  126314.00  126314.00  126314.00   
mean     1495.24    1495.24      41.71     102.73    1495.24    1495.24   
std       112.14     112.46      10.63      14.81     112.14     112.46   
min      1091.64    1085.77      10.15       0.00    1091.64    1085.77   
25%      1417.24    1416.99      34.10      93.00    1417.24    1416.99   
50%      1500.95    1500.95      42.11     103.00    1500.95    1500.95   
75%      1576.06    1576.29      49.64     112.00    1576.06    1576.29   
max      1853.10    1853.10      71.11     186.00    1853.10    1853.10   

        forecast  
count  126314.00  
mean        0.50  
std         0.22  
min         0.02  
25%         0.33  
50%         0.50  
75%         0.67  
max         0.98

This function show you some more basic descriptive statisitics for all numerical columns. 

NOTE: .describe() only analyzes numerical columns by default. But you can provide other data types  if you use the "include" parameter

In [17]:
import numpy as np
nba.describe(include=np.object)

game_id   lg_id date_game team_id fran_id  opp_id opp_fran  \
count         126314  126314    126314  126314  126314  126314   126314   
unique         63157       2     12426     104      53     104       53   
top     201211020ATL     NBA  1/2/2009     BOS  Lakers     BOS   Lakers   
freq               2  118016        30    5997    6024    5997     6024   

       game_location game_result           notes  
count         126314      126314            5424  
unique             3           2             231  
top                A           W  at New York NY  
freq           63138       63157             440

.describe wont try to calculate a *MEAN or Standard Deviation* for the object columns, since they mostly include text strings but still will provide some descriptivie statistics. 

Take a look at the team_id and Fran_id columns. the data set contains 104 team IDs but only 53 franchise IDs. Furthermore, the most frequent team ID is BOS (Boston), and the most frequent franchise is lakers.   How is this possible?

Lets explore a bit more...

##**Exploratory Data Analysis**

exploratatory data analysis can help you answer questions about your dataset. for example, you can examine how often values occur in a specific column

In [18]:
nba["team_id"].value_counts()

BOS    5997
NYK    5769
LAL    5078
DET    4985
PHI    4533
       ... 
PIT      60
TRH      60
INJ      60
DTF      60
SDS      11
Name: team_id, Length: 104, dtype: int64

In [19]:
nba["fran_id"].value_counts()

Lakers          6024
Celtics         5997
Knicks          5769
Warriors        5657
Pistons         5650
Sixers          5644
Hawks           5572
Kings           5475
Wizards         4582
Spurs           4309
Bulls           4307
Pacers          4227
Thunder         4178
Rockets         4154
Nuggets         4120
Nets            4106
Suns            4080
Bucks           4034
Trailblazers    3870
Cavaliers       3810
Clippers        3733
Jazz            3555
Mavericks       3013
Heat            2371
Pelicans        2254
Magic           2207
Timberwolves    2131
Grizzlies       1657
Raptors         1634
Hornets          894
Colonels         846
Squires          799
Spirits          777
Stars            756
Sounds           697
Baltimore        467
Floridians       440
Condors          430
Capitols         291
Olympians        282
Sails            274
Stags            260
Bombers          249
Steamrollers     168
Packers           72
Redskins          65
Rebels            63
Denver       

It seems that a team named "lakers" played 6024 games but only 5078 of those were played by the Los Angeles lakers.  who is the other lakers team?

In [21]:
nba.loc[nba["fran_id"] == "Lakers", "team_id"].value_counts()

LAL    5078
MNL     946
Name: team_id, dtype: int64

Las Angeles lakers (LAL) and Minneapolis Lakers (MNL) played 946 games. We can even find out when those games were played. 

In [22]:
nba.loc[nba["team_id"] == "MNL", "date_game"].min()

'1/1/1949'

In [23]:
nba.loc[nba["team_id"] == "MNL", "date_game"].max()

'4/9/1959'

In [24]:
nba.loc[nba["team_id"] == "MNL", "date_game"].agg(("min", "max"))

min    1/1/1949
max    4/9/1959
Name: date_game, dtype: object

Looks like the Mineanapolis lakers played between the years 1949 and 1959. This explains why we didnt recognize the team. Youve also found out why the Boston Celtics (BOS)played the most games in the dataset.  Lets analyze their history also a little bit.  Let find out how many points the boston celtics scored during all matches contained in the dataset. 

In [25]:
nba.loc[nba["team_id"] == "BOS", "pts"].sum()

626484

The Boston Celtics scored a total of 626,484 points

#**Pandas Data strucutures**

###Understanding Series Objects

Pyhons most basic data structure is a list. Which is also a great starting point for getting to know pandas.series Objects. 

In [0]:
revenues = pd.Series([5555, 7000, 1980])

In [29]:
revenues

0    5555
1    7000
2    1980
dtype: int64

you've used the list [5555, 7000, 1980] to create a series object called revenues. a series object wraps two components:

1.   A sequence of values
2.   A sequence of identifiers, which is the index

you can acces these components with .values and .index, respectively



In [31]:
revenues.values

array([5555, 7000, 1980])

In [32]:
revenues.index

RangeIndex(start=0, stop=3, step=1)

revenue.values returns the values in a series, revenues.index returns the positional index

NOTE: It may be of interest to note that the values of a series object are actually N-Dimensional arrays.  See NumPy for more details (https://numpy.org/)